mypy/expandtype.py

```diff
@@ -226,6 +227,14 @@ def visit_type_var(self, t: TypeVarType) -> Type:
             # TODO: do we really need to do this?
             # If I try to remove this special-casing ~40 tests fail on reveal_type().
             return repl.copy_modified(last_known_value=None)
+        if isinstance(repl, TypeVarType) and repl.has_default():
+            if (tvar_id := repl.id) in self.recursive_tvar_guard:
+                return self.recursive_tvar_guard[tvar_id] or repl
+            self.recursive_tvar_guard[tvar_id] = None
+            repl = repl.accept(self)
+            if isinstance(repl, TypeVarType):
+                repl.default = repl.default.accept(self)
```

Comment:
Shouldn't we `.copy_modified()` here?

Reply:
That doesn't work, unfortunately. This here is used to link nested TypeVarTypes so if it's changed / evaluated in one location the other stays in sync.

An example from the tests
```py
T1 = TypeVar("T1", default=str)
T2 = TypeVar("T2", default=T1)

class ClassD1(Generic[T1, T2]): ...

k = ClassD1()
reveal_type(k)  # should be `ClassD1[str, str]`
```

For the first pass the type variables are as follows
```py
T1`1 = str
T2`2 = T1`1 = str
```

The section here now replaces the default for `T2` with the tbd of the default of `T1`:
```py
T1`1 = str
T2`2 = <result of T1`1>
```

--
With `.copy_modified` it would be a copy of the `T1'1` thus when it's evaluated it's not used for `T2`. The test would still return
```py
ClassD1[str, T1`1 = str]
```

Reply:
Looked at it more closely. It's basically as  said above. In particular, it's called in `freshend_function_type_vars`. We enter this function with linked TypeVars
```py
callee.variables[0] == callee.variables[1].default  # True
```
Then create new ones for each and link them again in `expand_type`, so that
```py
tvs[0] == tvs[1].default  # True

# and by extension
fresh.variables[0] == fresh.variables[1].default  # True
```

https://github.com/python/mypy/blob/517f5aee23ba218f615bcd4427bca62f120bc222/mypy/expandtype.py#L119-L130